Error: versionId must not be blank

Fullstop should not try to load an application version from Kio, if the versionId is blank.

```
ERROR [pool-7-thread-1] o.z.s.f.PluginEventsProcessor - java.lang.IllegalArgumentException: versionId must not be blank
com.google.common.util.concurrent.UncheckedExecutionException: java.lang.IllegalArgumentException: versionId must not be blank
at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2203) ~[guava-18.0.jar!/:?]
at com.google.common.cache.LocalCache.get(LocalCache.java:3937) ~[guava-18.0.jar!/:?]
at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3941) ~[guava-18.0.jar!/:?]
at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4824) ~[guava-18.0.jar!/:?]
at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4830) ~[guava-18.0.jar!/:?]
at org.zalando.stups.fullstop.plugin.provider.impl.KioVersionProviderImpl.apply(KioVersionProviderImpl.java:62) ~[fullstop-instance-plugin-support-1.2.0-SNAPSHOT.jar!/:?]
at org.zalando.stups.fullstop.plugin.provider.impl.KioVersionProviderImpl.apply(KioVersionProviderImpl.java:21) ~[fullstop-instance-plugin-support-1.2.0-SNAPSHOT.jar!/:?]
at org.zalando.stups.fullstop.plugin.impl.EC2InstanceContextImpl.getKioVersion(EC2InstanceContextImpl.java:154) ~[fullstop-instance-plugin-support-1.2.0-SNAPSHOT.jar!/:?]
at org.zalando.stups.fullstop.plugin.ApplicationRegistryPlugin.process(ApplicationRegistryPlugin.java:73) ~[fullstop-registry-plugin-1.2.0-SNAPSHOT.jar!/:?]
at org.zalando.stups.fullstop.plugin.AbstractEC2InstancePlugin$$Lambda$56/296355274.accept(Unknown Source) ~[?:?]
at java.util.ArrayList.forEach(ArrayList.java:1249) ~[?:1.8.0_45-internal]
at org.zalando.stups.fullstop.plugin.AbstractEC2InstancePlugin.processEvent(AbstractEC2InstancePlugin.java:36) ~[fullstop-instance-plugin-support-1.2.0-SNAPSHOT.jar!/:?]
at org.zalando.stups.fullstop.PluginEventsProcessor.doProcess(PluginEventsProcessor.java:52) [fullstop-processing-1.2.0-SNAPSHOT.jar!/:?]
at org.zalando.stups.fullstop.PluginEventsProcessor.doProcess(PluginEventsProcessor.java:43) [fullstop-processing-1.2.0-SNAPSHOT.jar!/:?]
at org.zalando.stups.fullstop.PluginEventsProcessor$$Lambda$49/1107547513.accept(Unknown Source) [fullstop-processing-1.2.0-SNAPSHOT.jar!/:?]
at java.util.ArrayList.forEach(ArrayList.java:1249) [?:1.8.0_45-internal]
at org.zalando.stups.fullstop.PluginEventsProcessor.process(PluginEventsProcessor.java:33) [fullstop-processing-1.2.0-SNAPSHOT.jar!/:?]
at com.amazonaws.services.cloudtrail.processinglibrary.reader.EventReader.emitEvents(EventReader.java:229) [aws-cloudtrail-processing-library-1.0.2.jar!/:?]
at com.amazonaws.services.cloudtrail.processinglibrary.reader.EventReader.processSource(EventReader.java:156) [aws-cloudtrail-processing-library-1.0.2.jar!/:?]
at com.amazonaws.services.cloudtrail.processinglibrary.AWSCloudTrailProcessingExecutor$ScheduledJob$1.run(AWSCloudTrailProcessingExecutor.java:178) [aws-cloudtrail-processing-library-1.0.2.jar!/:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_45-internal]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_45-internal]
at java.lang.Thread.run(Thread.java:745) [?:1.8.0_45-internal]
Caused by: java.lang.IllegalArgumentException: versionId must not be blank
at org.springframework.util.Assert.hasText(Assert.java:168) ~[spring-core-4.2.3.RELEASE.jar!/:4.2.3.RELEASE]
at org.zalando.stups.clients.kio.spring.RestTemplateKioOperations.getApplicationVersion(RestTemplateKioOperations.java:132) ~[kio-client-java-spring-0.11.0.jar!/:?]
at org.zalando.stups.clients.kio.spring.RestTemplateKioOperations.getApplicationVersion(RestTemplateKioOperations.java:132) ~[kio-client-java-spring-0.11.0.jar!/:?]
at org.zalando.stups.fullstop.hystrix.HystrixKioOperations.getApplicationVersion(HystrixKioOperations.java:64) ~[fullstop.jar!/:?]
at sun.reflect.GeneratedMethodAccessor342.invoke(Unknown Source) ~[?:?]
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_45-internal]
at java.lang.reflect.Method.invoke(Method.java:497) ~[?:1.8.0_45-internal]
at com.netflix.hystrix.contrib.javanica.command.MethodExecutionAction.execute(MethodExecutionAction.java:116) ~[hystrix-javanica-1.4.21.jar!/:1.4.21]
at com.netflix.hystrix.contrib.javanica.command.MethodExecutionAction.executeWithArgs(MethodExecutionAction.java:89) ~[hystrix-javanica-1.4.21.jar!/:1.4.21]
at com.netflix.hystrix.contrib.javanica.command.MethodExecutionAction.execute(MethodExecutionAction.java:78) ~[hystrix-javanica-1.4.21.jar!/:1.4.21]
at com.netflix.hystrix.contrib.javanica.command.GenericCommand$1.execute(GenericCommand.java:47) ~[hystrix-javanica-1.4.21.jar!/:1.4.21]
at com.netflix.hystrix.contrib.javanica.command.AbstractHystrixCommand.process(AbstractHystrixCommand.java:158) ~[hystrix-javanica-1.4.21.jar!/:1.4.21]
at com.netflix.hystrix.contrib.javanica.command.GenericCommand.run(GenericCommand.java:44) ~[hystrix-javanica-1.4.21.jar!/:1.4.21]
at com.netflix.hystrix.HystrixCommand$1.call(HystrixCommand.java:294) ~[hystrix-core-1.4.10.jar!/:1.4.10]
at com.netflix.hystrix.HystrixCommand$1.call(HystrixCommand.java:289) ~[hystrix-core-1.4.10.jar!/:1.4.10]
at rx.Observable$1.call(Observable.java:144) ~[rxjava-1.0.10.jar!/:1.0.10]
at rx.Observable$1.call(Observable.java:136) ~[rxjava-1.0.10.jar!/:1.0.10]
at rx.Observable$1.call(Observable.java:144) ~[rxjava-1.0.10.jar!/:1.0.10]
at rx.Observable$1.call(Observable.java:136) ~[rxjava-1.0.10.jar!/:1.0.10]
at rx.Observable$1.call(Observable.java:144) ~[rxjava-1.0.10.jar!/:1.0.10]
at rx.Observable$1.call(Observable.java:136) ~[rxjava-1.0.10.jar!/:1.0.10]
at rx.Observable.unsafeSubscribe(Observable.java:7507) ~[rxjava-1.0.10.jar!/:1.0.10]
at com.netflix.hystrix.AbstractCommand$5.call(AbstractCommand.java:534) ~[hystrix-core-1.4.10.jar!/:1.4.10]
at com.netflix.hystrix.AbstractCommand$5.call(AbstractCommand.java:509) ~[hystrix-core-1.4.10.jar!/:1.4.10]
at rx.Observable.unsafeSubscribe(Observable.java:7507) ~[rxjava-1.0.10.jar!/:1.0.10]
at rx.internal.operators.OperatorSubscribeOn$1$1.call(OperatorSubscribeOn.java:62) ~[rxjava-1.0.10.jar!/:1.0.10]
at com.netflix.hystrix.strategy.concurrency.HystrixContexSchedulerAction$1.call(HystrixContexSchedulerAction.java:56) ~[hystrix-core-1.4.10.jar!/:1.4.10]
at com.netflix.hystrix.strategy.concurrency.HystrixContexSchedulerAction$1.call(HystrixContexSchedulerAction.java:47) ~[hystrix-core-1.4.10.jar!/:1.4.10]
at com.netflix.hystrix.strategy.concurrency.HystrixContexSchedulerAction.call(HystrixContexSchedulerAction.java:69) ~[hystrix-core-1.4.10.jar!/:1.4.10]
at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55) ~[rxjava-1.0.10.jar!/:1.0.10]
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_45-internal]
at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_45-internal]
... 3 more
```

--- a/fullstop-plugins/fullstop-instance-plugin-support/pom.xml
+++ b/fullstop-plugins/fullstop-instance-plugin-support/pom.xml
@@ -51,6 +51,11 @@
             <artifactId>snakeyaml</artifactId>
         </dependency>
         <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>3.4</version>
+        </dependency>
+        <dependency>
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-test</artifactId>
             <scope>test</scope>

--- a/fullstop-plugins/fullstop-instance-plugin-support/src/main/java/org/zalando/stups/fullstop/plugin/impl/EC2InstanceContextImpl.java
+++ b/fullstop-plugins/fullstop-instance-plugin-support/src/main/java/org/zalando/stups/fullstop/plugin/impl/EC2InstanceContextImpl.java
@@ -5,6 +5,7 @@ import com.amazonaws.regions.Region;
 import com.amazonaws.services.cloudtrail.processinglibrary.model.CloudTrailEvent;
 import com.amazonaws.services.ec2.model.Image;
 import com.jayway.jsonpath.JsonPath;
+import org.apache.commons.lang3.StringUtils;
 import org.zalando.stups.clients.kio.Application;
 import org.zalando.stups.clients.kio.Approval;
 import org.zalando.stups.clients.kio.Version;
@@ -126,22 +127,22 @@ public class EC2InstanceContextImpl implements EC2InstanceContext {
 
     @Override
     public Optional<String> getApplicationId() {
-        return getTaupageYaml().map(data -> (String) data.get("application_id"));
+        return getTaupageYaml().map(data -> (String) data.get("application_id")).map(StringUtils::trimToNull);
     }
 
     @Override
     public Optional<String> getVersionId() {
-        return getTaupageYaml().map(data -> (String) data.get("application_version"));
+        return getTaupageYaml().map(data -> (String) data.get("application_version")).map(StringUtils::trimToNull);
     }
 
     @Override
     public Optional<String> getSource() {
-        return getTaupageYaml().map(data -> (String) data.get("source"));
+        return getTaupageYaml().map(data -> (String) data.get("source")).map(StringUtils::trimToNull);
     }
 
     @Override
     public Optional<String> getRuntime() {
-        return getTaupageYaml().map(m -> (String) m.get("runtime"));
+        return getTaupageYaml().map(m -> (String) m.get("runtime")).map(StringUtils::trimToNull);
     }
 
     @Override

--- a/fullstop-plugins/fullstop-instance-plugin-support/src/main/java/org/zalando/stups/fullstop/plugin/provider/impl/AmiIdProviderImpl.java
+++ b/fullstop-plugins/fullstop-instance-plugin-support/src/main/java/org/zalando/stups/fullstop/plugin/provider/impl/AmiIdProviderImpl.java
@@ -10,6 +10,7 @@ import com.google.common.cache.CacheLoader;
 import com.google.common.cache.LoadingCache;
 import com.jayway.jsonpath.JsonPath;
 import com.jayway.jsonpath.JsonPathException;
+import org.apache.commons.lang3.StringUtils;
 import org.slf4j.Logger;
 import org.zalando.stups.fullstop.plugin.EC2InstanceContext;
 import org.zalando.stups.fullstop.plugin.provider.AmiIdProvider;
@@ -34,7 +35,7 @@ public class AmiIdProviderImpl implements AmiIdProvider {
             .build(new CacheLoader<EC2InstanceContext, Optional<String>>() {
                 @Override
                 public Optional<String> load(@Nonnull EC2InstanceContext context) throws Exception {
-                    final Optional<String> amiId = getAmiId(context);
+                    final Optional<String> amiId = getAmiId(context).map(StringUtils::trimToNull);
                     if (!amiId.isPresent()) {
                         log.warn("No AMI id found for {}", context);
                     }

--- a/fullstop-plugins/fullstop-instance-plugin-support/src/main/java/org/zalando/stups/fullstop/plugin/provider/impl/KioVersionProviderImpl.java
+++ b/fullstop-plugins/fullstop-instance-plugin-support/src/main/java/org/zalando/stups/fullstop/plugin/provider/impl/KioVersionProviderImpl.java
@@ -3,7 +3,6 @@ package org.zalando.stups.fullstop.plugin.provider.impl;
 import com.google.common.cache.CacheBuilder;
 import com.google.common.cache.CacheLoader;
 import com.google.common.cache.LoadingCache;
-import org.apache.commons.lang3.StringUtils;
 import org.slf4j.Logger;
 import org.zalando.stups.clients.kio.KioOperations;
 import org.zalando.stups.clients.kio.NotFoundException;
@@ -47,10 +46,7 @@ public class KioVersionProviderImpl implements KioVersionProvider {
         final Optional<String> applicationId = context.getApplicationId();
         final Optional<String> versionId = context.getVersionId();
 
-        if (applicationId.isPresent() &&
-                versionId.isPresent() &&
-                StringUtils.trimToNull(applicationId.get()) != null &&
-                StringUtils.trimToNull(versionId.get()) != null) {
+        if (applicationId.isPresent() && versionId.isPresent()) {
             try {
                 return ofNullable(kioOperations.getApplicationVersion(applicationId.get(), versionId.get()));
             } catch (NotFoundException ignored) {

--- a/fullstop-plugins/fullstop-instance-plugin-support/src/main/java/org/zalando/stups/fullstop/plugin/provider/impl/KioVersionProviderImpl.java
+++ b/fullstop-plugins/fullstop-instance-plugin-support/src/main/java/org/zalando/stups/fullstop/plugin/provider/impl/KioVersionProviderImpl.java
@@ -3,6 +3,7 @@ package org.zalando.stups.fullstop.plugin.provider.impl;
 import com.google.common.cache.CacheBuilder;
 import com.google.common.cache.CacheLoader;
 import com.google.common.cache.LoadingCache;
+import org.apache.commons.lang3.StringUtils;
 import org.slf4j.Logger;
 import org.zalando.stups.clients.kio.KioOperations;
 import org.zalando.stups.clients.kio.NotFoundException;
@@ -46,7 +47,10 @@ public class KioVersionProviderImpl implements KioVersionProvider {
         final Optional<String> applicationId = context.getApplicationId();
         final Optional<String> versionId = context.getVersionId();
 
-        if (applicationId.isPresent() && versionId.isPresent()) {
+        if (applicationId.isPresent() &&
+                versionId.isPresent() &&
+                StringUtils.trimToNull(applicationId.get()) != null &&
+                StringUtils.trimToNull(versionId.get()) != null) {
             try {
                 return ofNullable(kioOperations.getApplicationVersion(applicationId.get(), versionId.get()));
             } catch (NotFoundException ignored) {
@@ -61,5 +65,5 @@ public class KioVersionProviderImpl implements KioVersionProvider {
     public Optional<Version> apply(EC2InstanceContext context) {
         return cache.getUnchecked(context);
     }
-    
+
 }

--- a/fullstop-plugins/fullstop-instance-plugin-support/src/test/java/org/zalando/stups/fullstop/plugin/provider/impl/KioVersionProviderImplTest.java
+++ b/fullstop-plugins/fullstop-instance-plugin-support/src/test/java/org/zalando/stups/fullstop/plugin/provider/impl/KioVersionProviderImplTest.java
@@ -76,30 +76,6 @@ public class KioVersionProviderImplTest {
     }
 
     @Test
-    public void testApplicationIdBlank() throws Exception {
-        when(ec2InstanceContextMock.getApplicationId()).thenReturn(Optional.of(" "));
-        when(ec2InstanceContextMock.getVersionId()).thenReturn(Optional.of(VERSION_ID));
-
-        Optional<Version> result = kioVersionProvider.apply(ec2InstanceContextMock);
-        assertThat(result).isEmpty();
-
-        verify(ec2InstanceContextMock).getApplicationId();
-        verify(ec2InstanceContextMock).getVersionId();
-    }
-
-    @Test
-    public void testVersionIdBlank() throws Exception {
-        when(ec2InstanceContextMock.getApplicationId()).thenReturn(Optional.of(INSTANCE_ID));
-        when(ec2InstanceContextMock.getVersionId()).thenReturn(Optional.of(" "));
-
-        Optional<Version> result = kioVersionProvider.apply(ec2InstanceContextMock);
-        assertThat(result).isEmpty();
-
-        verify(ec2InstanceContextMock).getApplicationId();
-        verify(ec2InstanceContextMock).getVersionId();
-    }
-
-    @Test
     public void testNotFoundException() throws Exception {
         when(ec2InstanceContextMock.getApplicationId()).thenReturn(Optional.of(INSTANCE_ID));
         when(ec2InstanceContextMock.getVersionId()).thenReturn(Optional.of(VERSION_ID));

--- a/fullstop-plugins/fullstop-instance-plugin-support/src/test/java/org/zalando/stups/fullstop/plugin/provider/impl/KioVersionProviderImplTest.java
+++ b/fullstop-plugins/fullstop-instance-plugin-support/src/test/java/org/zalando/stups/fullstop/plugin/provider/impl/KioVersionProviderImplTest.java
@@ -77,8 +77,8 @@ public class KioVersionProviderImplTest {
 
     @Test
     public void testApplicationIdBlank() throws Exception {
-        when(ec2InstanceContextMock.getApplicationId()).thenReturn(Optional.of(INSTANCE_ID));
-        when(ec2InstanceContextMock.getVersionId()).thenReturn(Optional.of(" "));
+        when(ec2InstanceContextMock.getApplicationId()).thenReturn(Optional.of(" "));
+        when(ec2InstanceContextMock.getVersionId()).thenReturn(Optional.of(VERSION_ID));
 
         Optional<Version> result = kioVersionProvider.apply(ec2InstanceContextMock);
         assertThat(result).isEmpty();

--- a/fullstop-plugins/fullstop-instance-plugin-support/src/test/java/org/zalando/stups/fullstop/plugin/provider/impl/KioVersionProviderImplTest.java
+++ b/fullstop-plugins/fullstop-instance-plugin-support/src/test/java/org/zalando/stups/fullstop/plugin/provider/impl/KioVersionProviderImplTest.java
@@ -3,14 +3,12 @@ package org.zalando.stups.fullstop.plugin.provider.impl;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
-import org.zalando.stups.clients.kio.Approval;
 import org.zalando.stups.clients.kio.KioOperations;
 import org.zalando.stups.clients.kio.NotFoundException;
 import org.zalando.stups.clients.kio.Version;
 import org.zalando.stups.fullstop.plugin.EC2InstanceContext;
 import org.zalando.stups.fullstop.plugin.provider.KioVersionProvider;
 
-import java.util.List;
 import java.util.Optional;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -69,6 +67,30 @@ public class KioVersionProviderImplTest {
     public void testVersionNotFound2() throws Exception {
         when(ec2InstanceContextMock.getApplicationId()).thenReturn(Optional.of(INSTANCE_ID));
         when(ec2InstanceContextMock.getVersionId()).thenReturn(Optional.empty());
+
+        Optional<Version> result = kioVersionProvider.apply(ec2InstanceContextMock);
+        assertThat(result).isEmpty();
+
+        verify(ec2InstanceContextMock).getApplicationId();
+        verify(ec2InstanceContextMock).getVersionId();
+    }
+
+    @Test
+    public void testApplicationIdBlank() throws Exception {
+        when(ec2InstanceContextMock.getApplicationId()).thenReturn(Optional.of(INSTANCE_ID));
+        when(ec2InstanceContextMock.getVersionId()).thenReturn(Optional.of(" "));
+
+        Optional<Version> result = kioVersionProvider.apply(ec2InstanceContextMock);
+        assertThat(result).isEmpty();
+
+        verify(ec2InstanceContextMock).getApplicationId();
+        verify(ec2InstanceContextMock).getVersionId();
+    }
+
+    @Test
+    public void testVersionIdBlank() throws Exception {
+        when(ec2InstanceContextMock.getApplicationId()).thenReturn(Optional.of(INSTANCE_ID));
+        when(ec2InstanceContextMock.getVersionId()).thenReturn(Optional.of(" "));
 
         Optional<Version> result = kioVersionProvider.apply(ec2InstanceContextMock);
         assertThat(result).isEmpty();

